cloudapi/insights/client.go

```diff
@@ -180,9 +202,75 @@ func dialOptionsFromClientConfig(cfg ClientConfig) ([]grpc.DialOption, error) {
 		opts = append(opts, grpc.WithPerRPCCredentials(newPerRPCCredentials(cfg.AuthConfig)))
 	}
 
+	rI, err := retryInterceptor(cfg.RetryConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create retry interceptors: %w", err)
+	}
+
+	opts = append(opts, grpc.WithChainUnaryInterceptor(rI...))
+
 	return opts, nil
 }
 
+func retryInterceptor(retryConfig ClientRetryConfig) ([]grpc.UnaryClientInterceptor, error) {
+	rSC, err := retryableStatusCodes(retryConfig.RetryableStatusCodes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse retryable status codes: %w", err)
+	}
+	withCodes := grpcRetry.WithCodes(rSC...)
+	withMax := grpcRetry.WithMax(retryConfig.MaxAttempts)
+	withPerRetryTimeout := grpcRetry.WithPerRetryTimeout(retryConfig.PerRetryTimeout)
+
+	backoffConfig := retryConfig.BackoffConfig
+	if backoffConfig.Enabled {
+		backoff := grpcRetry.WithBackoff(
+			grpcRetry.BackoffExponentialWithJitter(backoffConfig.WaitBetween, backoffConfig.JitterFraction))
+		unaryInterceptor := grpcRetry.UnaryClientInterceptor(withCodes, withMax, withPerRetryTimeout, backoff)
+		return []grpc.UnaryClientInterceptor{unaryInterceptor}, nil
+	}
+
+	unaryInterceptor := grpcRetry.UnaryClientInterceptor(withCodes, withMax, withPerRetryTimeout)
+	return []grpc.UnaryClientInterceptor{unaryInterceptor}, nil
+}
+
+func retryableStatusCodes(retryableStatusCodes string) ([]codes.Code, error) {
+	statusCodeMap := map[string]codes.Code{
+		"OK":                  codes.OK,
+		"CANCELLED":           codes.Canceled,
+		"UNKNOWN":             codes.Unknown,
+		"INVALID_ARGUMENT":    codes.InvalidArgument,
+		"DEADLINE_EXCEEDED":   codes.DeadlineExceeded,
+		"NOT_FOUND":           codes.NotFound,
+		"ALREADY_EXISTS":      codes.AlreadyExists,
+		"PERMISSION_DENIED":   codes.PermissionDenied,
+		"UNAUTHENTICATED":     codes.Unauthenticated,
+		"RESOURCE_EXHAUSTED":  codes.ResourceExhausted,
+		"FAILED_PRECONDITION": codes.FailedPrecondition,
+		"ABORTED":             codes.Aborted,
+		"OUT_OF_RANGE":        codes.OutOfRange,
+		"UNIMPLEMENTED":       codes.Unimplemented,
+		"INTERNAL":            codes.Internal,
+		"UNAVAILABLE":         codes.Unavailable,
+		"DATA_LOSS":           codes.DataLoss,
+	}
+
+	if len(retryableStatusCodes) == 0 {
+		return nil, fmt.Errorf("no retryable status codes provided")
+	}
+
+	statusCodes := strings.Split(retryableStatusCodes, ",")
+	errorCodes := make([]codes.Code, 0, len(statusCodes))
+	for _, code := range statusCodes {
+		errorCode, ok := statusCodeMap[code]
```

Comment:
I feel like we might be better of using `UnsmarshalJSON` https://github.com/grpc/grpc-go/blob/v1.56.1/codes/codes.go#L219 

This iwll make it so we don't have to figure out if a code was added or not. I for example was certain you have missed "Unauthenticated" ... but it is just not in the same place as it is in the codes' code ( :wink: )

Reply:
It also takes the name 

https://github.com/grpc/grpc-go/blob/5b67e5ea449ef0686a0c0b6de48cd4cb63e3db2a/codes/codes.go#L239

Reply:
Yes, you are totally right; I updated `retryableStatusCodes` to use `UnmarshalJSON`. Thanks!